power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp

```diff
@@ -6,16 +6,306 @@
+    std::vector<ComponentBuffer> component_buffers_; // list of components, then all scenario flatten
+    // msgpack pakcer
+    msgpack::sbuffer msgpack_buffer_{};
+    msgpack::packer<msgpack::sbuffer> packer_;
+    bool use_compact_list_{};
+    std::map<MetaComponent const*, std::vector<MetaAttribute const*>> attributes_;
+    // json
+    std::string json_buffer_;
+
+    void store_buffers(char const** components, Idx const* elements_per_scenario, Idx const** indptrs,
+                       void const** data) {
+        scenario_buffers_.resize(batch_size_);
+        for (Idx scenario = 0; scenario != batch_size_; ++scenario) {
+            scenario_buffers_[scenario] =
+                create_scenario_buffer_view(components, elements_per_scenario, indptrs, data, scenario);
+        }
+        component_buffers_ =
+            create_scenario_buffer_view(components, elements_per_scenario, indptrs, data).component_buffers;
+    }
+
+    ScenarioBuffer create_scenario_buffer_view(char const** components, Idx const* elements_per_scenario,
+                                               Idx const** indptrs, void const** data, Idx scenario = -1) {
+        ScenarioBuffer scenario_buffer{};
+        Idx const begin_scenario = scenario < 0 ? 0 : scenario;
+        Idx const end_scenario = scenario < 0 ? batch_size_ : begin_scenario + 1;
+        for (Idx component = 0; component != n_components_; ++component) {
+            ComponentBuffer component_buffer{};
+            component_buffer.component = &dataset_->get_component(components[component]);
+            if (elements_per_scenario[component] < 0) {
+                component_buffer.data =
+                    component_buffer.component->advance_ptr(data[component], indptrs[component][begin_scenario]);
+                component_buffer.size = indptrs[component][end_scenario] - indptrs[component][begin_scenario];
+            } else {
+                component_buffer.data = component_buffer.component->advance_ptr(
+                    data[component], elements_per_scenario[component] * begin_scenario);
+                component_buffer.size = elements_per_scenario[component] * (end_scenario - begin_scenario);
+            }
+            // only store the view if it is non-empty
+            if (component_buffer.size > 0) {
+                scenario_buffer.component_buffers.push_back(component_buffer);
+            }
+        }
+        return scenario_buffer;
+    }
+
+    void check_attributes() {
+        attributes_ = {};
+        for (auto const& buffer : component_buffers_) {
+            std::vector<MetaAttribute const*> attributes;
+            for (auto const& attribute : buffer.component->attributes) {
+                // if not all the values of an attribute are nan
+                // add this attribute to the list
+                if (!attribute.check_all_nan(buffer.data, buffer.size)) {
+                    attributes.push_back(&attribute);
+                }
+            }
```

Comment:
do you think we should also provide a sparse packer? (like with how we used to use the dependent datasets)

Reply:
What do you mean exactly? Sparse regarding different number of elements per scenario? That we already support.

Reply:
oh sorry i misunderstood. that was indeed what i meant